File: 0876-hand-of-straights/0876-hand-of-straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        int n = hand.length;
        if(n % groupSize != 0) return false;
        
        Arrays.sort(hand);

        Map<Integer, Integer> freq = new HashMap<>();
        for(int h: hand) {
            freq.put(h, freq.getOrDefault(h, 0) + 1);
        }
        
        int groups = 0;
        for(int i=0; i<n; i++) {
            if(freq.get(hand[i]) > 0) {
                freq.put(hand[i], freq.get(hand[i])-1);
                int size = 1;
                int next = hand[i] + 1;
                while(freq.containsKey(next) && freq.get(next) > 0 && size != groupSize) {
                    freq.put(next, freq.get(next) - 1);
                    size++;
                    next++;
                }

                if(size == groupSize) groups++;
            }
        }
        return groups == (n / groupSize);
    }
}

File: 0876-hand-of-straights/0876-hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n = hand.size();
        if(n % groupSize != 0) return false;
        
        unordered_map<int, int> freq;
        for(int h: hand) freq[h]++;

        sort(hand.begin(), hand.end());
        int groups = 0;
        for(int i=0; i<n; i++) {
            if(freq[hand[i]] > 0) {
                freq[hand[i]]--;
                int size = 1;
                int next = hand[i] + 1;
                while(freq.find(next) != freq.end() && freq[next] > 0 && size != groupSize) {
                    freq[next]--;
                    next++;
                    size++;
                }
                if(size = groupSize) groups++;
            }
        }

        return groups == (n/groupSize);
    }
};